README
Getting SDL to play nice on Droid 4
![sdl1.2 with opengl->gl4es](screenshots/lgs-ba-arx_libertatis.jpg "Arx Libertatis")
![Plain sdl1.2](screenshots/lgs-aa-openjazz.jpg "OpenJazz")

                         Simple DirectMedia Layer

                                  (SDL)

                                Version 1.2
			for Maemo-Leste / Droid4
---
http://www.libsdl.org/

This is the Simple DirectMedia Layer, a general API that provides low
level access to audio, keyboard, mouse, joystick, 3D hardware via OpenGL,
and 2D framebuffer across multiple platforms.

The current version supports Linux, Windows CE/95/98/ME/XP/Vista, BeOS,
MacOS Classic, Mac OS X, FreeBSD, NetBSD, OpenBSD, BSD/OS, Solaris, IRIX,
and QNX.  The code contains support for Dreamcast, Atari, AIX, OSF/Tru64,
RISC OS, SymbianOS, Nintendo DS, and OS/2, but these are not officially
supported.

SDL is written in C, but works with C++ natively, and has bindings to
several other languages, including Ada, C#, Eiffel, Erlang, Euphoria,
Guile, Haskell, Java, Lisp, Lua, ML, Objective C, Pascal, Perl, PHP,
Pike, Pliant, Python, Ruby, and Smalltalk.

This library is distributed under GNU LGPL version 2, which can be
found in the file  "COPYING".  This license allows you to use SDL
freely in commercial programs as long as you link with the dynamic
library.

The best way to learn how to use SDL is to check out the header files in
the "include" subdirectory and the programs in the "test" subdirectory.
The header files and test programs are well commented and always up to date.
More documentation is available in HTML format in "docs/index.html", and
a documentation wiki is available online at:
	http://www.libsdl.org/cgi/docwiki.cgi

The test programs in the "test" subdirectory are in the public domain.

Frequently asked questions are answered online:
	http://www.libsdl.org/faq.php

If you need help with the library, or just want to discuss SDL related
issues, you can join the developers mailing list:
	http://www.libsdl.org/mailing-list.php

Enjoy!
	Sam Lantinga				(slouken@libsdl.org)



debian/changelog
libsdl1.2 (1:1.2.15-7+2m7.1) unstable; urgency=medium

  * Increment build number

 -- Jenkins Auto Builder <spam1@wizzup.org>  Thu, 16 Jul 2020 11:33:46 +0100

libsdl1.2 (1:1.2.15-7) unstable; urgency=medium

  * Add rebased Maemo Fremantle patches.

 -- Ivan J. <parazyd@dyne.org>  Mon, 13 Jul 2020 11:25:49 +0200

libsdl1.2 (1.2.15+dfsg2-6) UNRELEASED; urgency=medium

  * Trim trailing whitespace.
  * Re-export upstream signing key without extra signatures.

 -- Debian Janitor <janitor@jelmer.uk>  Thu, 24 Oct 2019 09:14:10 +0000

libsdl1.2 (1.2.15+dfsg2-5) unstable; urgency=medium

  [ Abhijith PA ]
  * Fix CVE-2019-7572, CVE-2019-7573, CVE-2019-7574, CVE-2019-7575
    CVE-2019-7576, CVE-2019-7577, CVE-2019-7578, CVE-2019-7635, CVE-2019-7636
    CVE-2019-7637, CVE-2019-7638
    (Closes: #924609)

  [ Felix Geyer ]
  * Fix CVE-2019-13616

 -- Felix Geyer <fgeyer@debian.org>  Tue, 17 Sep 2019 22:34:12 +0200

libsdl1.2 (1.2.15+dfsg2-4) unstable; urgency=medium

  * d/rules: Add @ in 'tar --mtime="@$(SOURCE_DATE_EPOCH)"', otherwise the
    date is not valid

 -- Manuel A. Fernandez Montecelo <mafm@debian.org>  Thu, 25 Oct 2018 01:47:02 +0200

libsdl1.2 (1.2.15+dfsg2-3) unstable; urgency=medium

  * Bump Policy Standards-Version to 4.2.1 (no changes needed)
  * d/control
    - Use https in Homepage URL
    - Set "Rules-Requires-Root: no"
    - Drop ancient build-dependency on "dpkg-dev (>= 1.16.1~)"
  * d/watch: use https in URL
  * d/copyright: use https in Format, Source and Upstream-Contact URLs
  * Switch to debhelper compat level v11
    - Drop "--with autoreconf --parallel", they are the default now, and
      build-dependency on dh-autoreconf
    - Drop override of dh_installexamples and instead use file
      debian/libsdl1.2-dev.examples
  * d/rules:
    - include /usr/share/dpkg/architecture.mk and .../pkg-info.mk to
      provide DEB_HOST_* and SOURCE_DATE_EPOCH, instead of getting them
      through shell invocations.
    - dh_install --fail-missing is deprecated and about to be removed,
      use dh_missing instead

 -- Manuel A. Fernandez Montecelo <mafm@debian.org>  Sat, 20 Oct 2018 14:25:59 +0200

libsdl1.2 (1.2.15+dfsg2-2) unstable; urgency=medium

  [ Gianfranco Costamagna ]
  * Team upload

  [ Ryan C. Gordon (icculus) ]
  * Disable --disable-loadso switch, it was useless and it is wrong
    in many cases (see LP: #1740517 for discussion)

 -- Gianfranco Costamagna <locutusofborg@debian.org>  Fri, 14 Sep 2018 19:52:34 +0200

libsdl1.2 (1.2.15+dfsg2-1) unstable; urgency=medium

  * Team upload
  * Ack previous NMU, thanks
  * debian/patches/SDL-1.2.15-vec_perm-ppc64le.patch:
    - fix test failures on ppc64el, by fixing a wrong ordering on BE
      systems. Thanks Frédéric Bonnard for the patch!

 -- Gianfranco Costamagna <locutusofborg@debian.org>  Wed, 30 May 2018 15:27:49 +0200

libsdl1.2 (1.2.15+dfsg2-0.1) unstable; urgency=medium

  * Non-maintainer upload.

  [ Manuel A. Fernandez Montecelo ]
  * Do not build-depend on libaudiofile, unused.
    Closes: #857670

  [ Jonas Smedegaard ]
  * Add script to check copyright and licensing of source.
  * Update copyright info:
    + Rewrite copyright file in machine-readable format.
    + Fix cover many secondary copyright holders and licenses.
  * Repackage upstream source to exclude more potentially non-DFSG code,
    and add patch to avoid missing code.
    Closes: Bug#834204.
  * Update watch file:
    + Bump to version 4.
    + Add usage comment.
    + Set repacksuffix.
  * Drop repackaging hints from git-buildpackage config: Handled by
    uscan + machine-readable copyright file.
  * Add lintian overrides regarding license in License-Reference field.
    See bug#786450.
  * Add lintian overrides regarding license name used both alone an in
    composition.

 -- Jonas Smedegaard <dr@jones.dk>  Tue, 18 Jul 2017 13:46:21 +0200

libsdl1.2 (1.2.15+dfsg1-4) unstable; urgency=high

  * Urgency "high" to fix problems with migration of -dbg package.  Thanks
    Yann Dirson.
  * Ensure proper migration from -dbg package using "dh_strip
    --dbgsym-migration"
  * Bump debhelper dependency for --dbgsym-migration
  * Fix for reproducible builds:
    - sort files by name when creating "examples.tar.gz"

 -- Manuel A. Fernandez Montecelo <mafm@debian.org>  Thu, 17 Mar 2016 12:15:31 +0000

libsdl1.2 (1.2.15+dfsg1-3) unstable; urgency=medium

  * Disable suppot for DirectFB (Closes: #816125)
  * d/control: Update Vcs-* URLs, use https
  * Fixes for reproducible builds:
    - timestamps_in_tarball, use --clamp-mtime with changelog as date
    - timestamps_in_gzip_headers, export GZIP="-9n"
  * Drop -dbg package

 -- Manuel A. Fernandez Montecelo <mafm@debian.org>  Sat, 12 Mar 2016 00:36:47 +0000

libsdl1.2 (1.2.15+dfsg1-2) unstable; urgency=medium

  * d/control: remove "Multi-Arch: no" from -dev package, despite what
    lintian says there's an error processing the field (#768353) and the
    package got rejected

 -- Manuel A. Fernandez Montecelo <mafm@debian.org>  Sun, 21 Feb 2016 17:46:19 +0000

libsdl1.2 (1.2.15+dfsg1-1) unstable; urgency=medium

  * New upstream release
    - Repack without SDL_qsort.c due to licence problem
  * Use newer relicenced version for SDL_qsort.c (Closes: #814445)
    Thanks the anonymous reporter, Ben Hutchings for looking into it and
    get in contact with the original author, the original author Gareth
    McCaughan for prompt relicencing.
    - d/patches/replace-relicenced-SDL_qsort.patch
  * Bump Policy Standards-Version to 3.9.7 (no changes needed)
  * Set -dev package as "Multi-Arch: no" (quells lintian error:
    old-style-config-script-multiarch-path)
  * d/watch: Add dfsg and pgpsig mangling
  * d/upstream/signing-key.asc: Import upstream gpg key for uscan to
    verify the orig tarball

 -- Manuel A. Fernandez Montecelo <mafm@debian.org>  Sun, 21 Feb 2016 16:00:03 +0000

libsdl1.2 (1.2.15-12) unstable; urgency=medium


src/audio/SDL_wave.c
/*
    SDL - Simple DirectMedia Layer
    Copyright (C) 1997-2012 Sam Lantinga

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

    Sam Lantinga
    slouken@libsdl.org
*/
#include "SDL_config.h"

/* Microsoft WAVE file loading routines */

#include "SDL_audio.h"
#include "SDL_wave.h"


static int ReadChunk(SDL_RWops *src, Chunk *chunk);

struct MS_ADPCM_decodestate {
	Uint8 hPredictor;
	Uint16 iDelta;
	Sint16 iSamp1;
	Sint16 iSamp2;
};
static struct MS_ADPCM_decoder {
	WaveFMT wavefmt;
	Uint16 wSamplesPerBlock;
	Uint16 wNumCoef;
	Sint16 aCoeff[7][2];
	/* * * */
	struct MS_ADPCM_decodestate state[2];
} MS_ADPCM_state;

static int InitMS_ADPCM(WaveFMT *format, int length)
{
	Uint8 *rogue_feel, *rogue_feel_end;
	int i;

	/* Set the rogue pointer to the MS_ADPCM specific data */
	if (length < sizeof(*format)) goto too_short;
	MS_ADPCM_state.wavefmt.encoding = SDL_SwapLE16(format->encoding);
	MS_ADPCM_state.wavefmt.channels = SDL_SwapLE16(format->channels);
	MS_ADPCM_state.wavefmt.frequency = SDL_SwapLE32(format->frequency);
	MS_ADPCM_state.wavefmt.byterate = SDL_SwapLE32(format->byterate);
	MS_ADPCM_state.wavefmt.blockalign = SDL_SwapLE16(format->blockalign);
	MS_ADPCM_state.wavefmt.bitspersample =
					 SDL_SwapLE16(format->bitspersample);
	rogue_feel = (Uint8 *)format+sizeof(*format);
	rogue_feel_end = (Uint8 *)format + length;
	if ( sizeof(*format) == 16 ) {
		rogue_feel += sizeof(Uint16);
	}
	if (rogue_feel + 4 > rogue_feel_end) goto too_short;
	MS_ADPCM_state.wSamplesPerBlock = ((rogue_feel[1]<<8)|rogue_feel[0]);
	rogue_feel += sizeof(Uint16);
	MS_ADPCM_state.wNumCoef = ((rogue_feel[1]<<8)|rogue_feel[0]);
	rogue_feel += sizeof(Uint16);
	if ( MS_ADPCM_state.wNumCoef != 7 ) {
		SDL_SetError("Unknown set of MS_ADPCM coefficients");
		return(-1);
	}
	for ( i=0; i<MS_ADPCM_state.wNumCoef; ++i ) {
		if (rogue_feel + 4 > rogue_feel_end) goto too_short;
		MS_ADPCM_state.aCoeff[i][0] = ((rogue_feel[1]<<8)|rogue_feel[0]);
		rogue_feel += sizeof(Uint16);
		MS_ADPCM_state.aCoeff[i][1] = ((rogue_feel[1]<<8)|rogue_feel[0]);
		rogue_feel += sizeof(Uint16);
	}
	return(0);
too_short:
	SDL_SetError("Unexpected length of a chunk with a MS ADPCM format");
	return(-1);
}

static Sint32 MS_ADPCM_nibble(struct MS_ADPCM_decodestate *state,
					Uint8 nybble, Sint16 *coeff)
{
	const Sint32 max_audioval = ((1<<(16-1))-1);
	const Sint32 min_audioval = -(1<<(16-1));
	const Sint32 adaptive[] = {
		230, 230, 230, 230, 307, 409, 512, 614,
		768, 614, 512, 409, 307, 230, 230, 230
	};
	Sint32 new_sample, delta;

	new_sample = ((state->iSamp1 * coeff[0]) +
		      (state->iSamp2 * coeff[1]))/256;
	if ( nybble & 0x08 ) {
		new_sample += state->iDelta * (nybble-0x10);
	} else {
		new_sample += state->iDelta * nybble;
	}
	if ( new_sample < min_audioval ) {
		new_sample = min_audioval;
	} else
	if ( new_sample > max_audioval ) {
		new_sample = max_audioval;
	}
	delta = ((Sint32)state->iDelta * adaptive[nybble])/256;
	if ( delta < 16 ) {
		delta = 16;
	}
	state->iDelta = (Uint16)delta;
	state->iSamp2 = state->iSamp1;
	state->iSamp1 = (Sint16)new_sample;
	return(new_sample);
}

static int MS_ADPCM_decode(Uint8 **audio_buf, Uint32 *audio_len)
{
	struct MS_ADPCM_decodestate *state[2];
	Uint8 *freeable, *encoded, *encoded_end, *decoded, *decoded_end;
	Sint32 encoded_len, samplesleft;
	Sint8 nybble, stereo;
	Sint16 *coeff[2];
	Sint32 new_sample;

	/* Allocate the proper sized output buffer */
	encoded_len = *audio_len;
	encoded = *audio_buf;
	encoded_end = encoded + encoded_len;
	freeable = *audio_buf;
	*audio_len = (encoded_len/MS_ADPCM_state.wavefmt.blockalign) * 
				MS_ADPCM_state.wSamplesPerBlock*
				MS_ADPCM_state.wavefmt.channels*sizeof(Sint16);
	*audio_buf = (Uint8 *)SDL_malloc(*audio_len);
	if ( *audio_buf == NULL ) {
		SDL_Error(SDL_ENOMEM);
		return(-1);
	}
	decoded = *audio_buf;
	decoded_end = decoded + *audio_len;

	/* Get ready... Go! */
	stereo = (MS_ADPCM_state.wavefmt.channels == 2);
	state[0] = &MS_ADPCM_state.state[0];
	state[1] = &MS_ADPCM_state.state[stereo];
	while ( encoded_len >= MS_ADPCM_state.wavefmt.blockalign ) {
		/* Grab the initial information for this block */
		if (encoded + 7 + (stereo ? 7 : 0) > encoded_end) goto invalid_size;
		state[0]->hPredictor = *encoded++;
		if ( stereo ) {
			state[1]->hPredictor = *encoded++;
		}
		if (state[0]->hPredictor >= 7 || state[1]->hPredictor >= 7) {
			goto invalid_predictor;
		}
		state[0]->iDelta = ((encoded[1]<<8)|encoded[0]);
		encoded += sizeof(Sint16);
		if ( stereo ) {
			state[1]->iDelta = ((encoded[1]<<8)|encoded[0]);
			encoded += sizeof(Sint16);
		}
		state[0]->iSamp1 = ((encoded[1]<<8)|encoded[0]);
		encoded += sizeof(Sint16);
		if ( stereo ) {
			state[1]->iSamp1 = ((encoded[1]<<8)|encoded[0]);
			encoded += sizeof(Sint16);
		}
		state[0]->iSamp2 = ((encoded[1]<<8)|encoded[0]);
		encoded += sizeof(Sint16);
		if ( stereo ) {
			state[1]->iSamp2 = ((encoded[1]<<8)|encoded[0]);
			encoded += sizeof(Sint16);
		}
		coeff[0] = MS_ADPCM_state.aCoeff[state[0]->hPredictor];
		coeff[1] = MS_ADPCM_state.aCoeff[state[1]->hPredictor];

		/* Store the two initial samples we start with */
		if (decoded + 4 + (stereo ? 4 : 0) > decoded_end) goto invalid_size;
		decoded[0] = state[0]->iSamp2&0xFF;
		decoded[1] = state[0]->iSamp2>>8;
		decoded += 2;
		if ( stereo ) {
			decoded[0] = state[1]->iSamp2&0xFF;
			decoded[1] = state[1]->iSamp2>>8;
			decoded += 2;
		}
		decoded[0] = state[0]->iSamp1&0xFF;
		decoded[1] = state[0]->iSamp1>>8;
		decoded += 2;
		if ( stereo ) {
			decoded[0] = state[1]->iSamp1&0xFF;
			decoded[1] = state[1]->iSamp1>>8;
			decoded += 2;
		}

		/* Decode and store the other samples in this block */
		samplesleft = (MS_ADPCM_state.wSamplesPerBlock-2)*
					MS_ADPCM_state.wavefmt.channels;
		while ( samplesleft > 0 ) {
			if (encoded + 1 > encoded_end) goto invalid_size;
			if (decoded + 4 > decoded_end) goto invalid_size;

			nybble = (*encoded)>>4;
			new_sample = MS_ADPCM_nibble(state[0],nybble,coeff[0]);
			decoded[0] = new_sample&0xFF;
			new_sample >>= 8;
			decoded[1] = new_sample&0xFF;
			decoded += 2;

			nybble = (*encoded)&0x0F;
			new_sample = MS_ADPCM_nibble(state[1],nybble,coeff[1]);
			decoded[0] = new_sample&0xFF;
			new_sample >>= 8;
			decoded[1] = new_sample&0xFF;
			decoded += 2;

			++encoded;
			samplesleft -= 2;
		}
		encoded_len -= MS_ADPCM_state.wavefmt.blockalign;
	}
	SDL_free(freeable);
	return(0);
invalid_size:
	SDL_SetError("Unexpected chunk length for a MS ADPCM decoder");
	SDL_free(freeable);
	return(-1);
invalid_predictor:
	SDL_SetError("Invalid predictor value for a MS ADPCM decoder");
	SDL_free(freeable);
	return(-1);
}

struct IMA_ADPCM_decodestate {
	Sint32 sample;
	Sint8 index;
};
static struct IMA_ADPCM_decoder {
	WaveFMT wavefmt;
	Uint16 wSamplesPerBlock;
	/* * * */
	struct IMA_ADPCM_decodestate state[2];
} IMA_ADPCM_state;

static int InitIMA_ADPCM(WaveFMT *format, int length)
{
	Uint8 *rogue_feel, *rogue_feel_end;

	/* Set the rogue pointer to the IMA_ADPCM specific data */
	if (length < sizeof(*format)) goto too_short;
	IMA_ADPCM_state.wavefmt.encoding = SDL_SwapLE16(format->encoding);
	IMA_ADPCM_state.wavefmt.channels = SDL_SwapLE16(format->channels);
	IMA_ADPCM_state.wavefmt.frequency = SDL_SwapLE32(format->frequency);
	IMA_ADPCM_state.wavefmt.byterate = SDL_SwapLE32(format->byterate);
	IMA_ADPCM_state.wavefmt.blockalign = SDL_SwapLE16(format->blockalign);
	IMA_ADPCM_state.wavefmt.bitspersample =
					 SDL_SwapLE16(format->bitspersample);
	rogue_feel = (Uint8 *)format+sizeof(*format);
	rogue_feel_end = (Uint8 *)format + length;
	if ( sizeof(*format) == 16 ) {
		rogue_feel += sizeof(Uint16);
	}
	if (rogue_feel + 2 > rogue_feel_end) goto too_short;
	IMA_ADPCM_state.wSamplesPerBlock = ((rogue_feel[1]<<8)|rogue_feel[0]);
	return(0);
too_short:
	SDL_SetError("Unexpected length of a chunk with an IMA ADPCM format");
	return(-1);
}

static Sint32 IMA_ADPCM_nibble(struct IMA_ADPCM_decodestate *state,Uint8 nybble)
{
	const Sint32 max_audioval = ((1<<(16-1))-1);
	const Sint32 min_audioval = -(1<<(16-1));
	const int index_table[16] = {
		-1, -1, -1, -1,
		 2,  4,  6,  8,
		-1, -1, -1, -1,
		 2,  4,  6,  8
	};
	const Sint32 step_table[89] = {
		7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 19, 21, 23, 25, 28, 31,
		34, 37, 41, 45, 50, 55, 60, 66, 73, 80, 88, 97, 107, 118, 130,
		143, 157, 173, 190, 209, 230, 253, 279, 307, 337, 371, 408,
		449, 494, 544, 598, 658, 724, 796, 876, 963, 1060, 1166, 1282,
		1411, 1552, 1707, 1878, 2066, 2272, 2499, 2749, 3024, 3327,
		3660, 4026, 4428, 4871, 5358, 5894, 6484, 7132, 7845, 8630,
		9493, 10442, 11487, 12635, 13899, 15289, 16818, 18500, 20350,
		22385, 24623, 27086, 29794, 32767
	};
	Sint32 delta, step;

	/* Clamp index value. The inital value can be invalid. */
	if ( state->index > 88 ) {
		state->index = 88;
	} else
	if ( state->index < 0 ) {
		state->index = 0;
	}

	/* Compute difference and new sample value */
	step = step_table[state->index];
	delta = step >> 3;
	if ( nybble & 0x04 ) delta += step;
	if ( nybble & 0x02 ) delta += (step >> 1);
	if ( nybble & 0x01 ) delta += (step >> 2);
	if ( nybble & 0x08 ) delta = -delta;
	state->sample += delta;

	/* Update index value */
	state->index += index_table[nybble];

	/* Clamp output sample */
	if ( state->sample > max_audioval ) {
		state->sample = max_audioval;
	} else
	if ( state->sample < min_audioval ) {
		state->sample = min_audioval;
	}
	return(state->sample);
}

/* Fill the decode buffer with a channel block of data (8 samples) */
static void Fill_IMA_ADPCM_block(Uint8 *decoded, Uint8 *encoded,
	int channel, int numchannels, struct IMA_ADPCM_decodestate *state)
{
	int i;
	Sint8 nybble;
	Sint32 new_sample;

	decoded += (channel * 2);
	for ( i=0; i<4; ++i ) {
		nybble = (*encoded)&0x0F;
		new_sample = IMA_ADPCM_nibble(state, nybble);
		decoded[0] = new_sample&0xFF;
		new_sample >>= 8;
		decoded[1] = new_sample&0xFF;
		decoded += 2 * numchannels;

		nybble = (*encoded)>>4;
		new_sample = IMA_ADPCM_nibble(state, nybble);
		decoded[0] = new_sample&0xFF;
		new_sample >>= 8;
		decoded[1] = new_sample&0xFF;
		decoded += 2 * numchannels;

		++encoded;
	}
}

static int IMA_ADPCM_decode(Uint8 **audio_buf, Uint32 *audio_len)
{
	struct IMA_ADPCM_decodestate *state;
	Uint8 *freeable, *encoded, *encoded_end, *decoded, *decoded_end;
	Sint32 encoded_len, samplesleft;
	unsigned int c, channels;

	/* Check to make sure we have enough variables in the state array */
	channels = IMA_ADPCM_state.wavefmt.channels;
	if ( channels > SDL_arraysize(IMA_ADPCM_state.state) ) {
		SDL_SetError("IMA ADPCM decoder can only handle %d channels",
					SDL_arraysize(IMA_ADPCM_state.state));
		return(-1);
	}
	state = IMA_ADPCM_state.state;

	/* Allocate the proper sized output buffer */
	encoded_len = *audio_len;
	encoded = *audio_buf;
	encoded_end = encoded + encoded_len;
	freeable = *audio_buf;
	*audio_len = (encoded_len/IMA_ADPCM_state.wavefmt.blockalign) * 
				IMA_ADPCM_state.wSamplesPerBlock*
				IMA_ADPCM_state.wavefmt.channels*sizeof(Sint16);
	*audio_buf = (Uint8 *)SDL_malloc(*audio_len);
	if ( *audio_buf == NULL ) {
		SDL_Error(SDL_ENOMEM);
		return(-1);
	}
	decoded = *audio_buf;
	decoded_end = decoded + *audio_len;

	/* Get ready... Go! */
	while ( encoded_len >= IMA_ADPCM_state.wavefmt.blockalign ) {
		/* Grab the initial information for this block */
		for ( c=0; c<channels; ++c ) {
			if (encoded + 4 > encoded_end) goto invalid_size;
			/* Fill the state information for this block */
			state[c].sample = ((encoded[1]<<8)|encoded[0]);
			encoded += 2;
			if ( state[c].sample & 0x8000 ) {
				state[c].sample -= 0x10000;
			}
			state[c].index = *encoded++;
			/* Reserved byte in buffer header, should be 0 */
			if ( *encoded++ != 0 ) {
				/* Uh oh, corrupt data?  Buggy code? */;
			}

			/* Store the initial sample we start with */
			if (decoded + 2 > decoded_end) goto invalid_size;
			decoded[0] = (Uint8)(state[c].sample&0xFF);
			decoded[1] = (Uint8)(state[c].sample>>8);
			decoded += 2;
		}

		/* Decode and store the other samples in this block */
		samplesleft = (IMA_ADPCM_state.wSamplesPerBlock-1)*channels;
		while ( samplesleft > 0 ) {
			for ( c=0; c<channels; ++c ) {
                                if (encoded + 4 > encoded_end) goto invalid_size;
				if (decoded + 4 * 4 * channels > decoded_end)
                                              goto invalid_size;
				Fill_IMA_ADPCM_block(decoded, encoded,
						c, channels, &state[c]);
				encoded += 4;
				samplesleft -= 8;
			}
			decoded += (channels * 8 * 2);
		}
		encoded_len -= IMA_ADPCM_state.wavefmt.blockalign;
	}
	SDL_free(freeable);
	return(0);
	invalid_size:
	SDL_SetError("Unexpected chunk length for an IMA ADPCM decoder");
	SDL_free(freeable);
	return(-1);
}

SDL_AudioSpec * SDL_LoadWAV_RW (SDL_RWops *src, int freesrc,
		SDL_AudioSpec *spec, Uint8 **audio_buf, Uint32 *audio_len)
{
	int was_error;
	Chunk chunk;
	int lenread;
	int MS_ADPCM_encoded, IMA_ADPCM_encoded;
	int samplesize;

	/* WAV magic header */
	Uint32 RIFFchunk;
	Uint32 wavelen = 0;
	Uint32 WAVEmagic;
	Uint32 headerDiff = 0;

	/* FMT chunk */
	WaveFMT *format = NULL;

	/* Make sure we are passed a valid data source */
	was_error = 0;
	if ( src == NULL ) {
		was_error = 1;
		goto done;
	}
		
	/* Check the magic header */
	RIFFchunk	= SDL_ReadLE32(src);
	wavelen		= SDL_ReadLE32(src);
	if ( wavelen == WAVE ) { /* The RIFFchunk has already been read */
		WAVEmagic = wavelen;
		wavelen   = RIFFchunk;
		RIFFchunk = RIFF;
	} else {
		WAVEmagic = SDL_ReadLE32(src);
	}
	if ( (RIFFchunk != RIFF) || (WAVEmagic != WAVE) ) {
		SDL_SetError("Unrecognized file type (not WAVE)");
		was_error = 1;
		goto done;
	}
	headerDiff += sizeof(Uint32); /* for WAVE */

	/* Read the audio data format chunk */
	chunk.data = NULL;
	do {
		if ( chunk.data != NULL ) {
			SDL_free(chunk.data);
			chunk.data = NULL;
		}
		lenread = ReadChunk(src, &chunk);
		if ( lenread < 0 ) {
			was_error = 1;
			goto done;
		}
		/* 2 Uint32's for chunk header+len, plus the lenread */
		headerDiff += lenread + 2 * sizeof(Uint32);
	} while ( (chunk.magic == FACT) || (chunk.magic == LIST) );

	/* Decode the audio data format */
	format = (WaveFMT *)chunk.data;
	if ( chunk.magic != FMT ) {
		SDL_SetError("Complex WAVE files not supported");
		was_error = 1;
		goto done;
	}
	MS_ADPCM_encoded = IMA_ADPCM_encoded = 0;
	switch (SDL_SwapLE16(format->encoding)) {
		case PCM_CODE:
			/* We can understand this */
			break;
		case MS_ADPCM_CODE:
			/* Try to understand this */
			if ( InitMS_ADPCM(format, lenread) < 0 ) {
				was_error = 1;
				goto done;
			}
			MS_ADPCM_encoded = 1;
			break;
		case IMA_ADPCM_CODE:
			/* Try to understand this */
			if ( InitIMA_ADPCM(format, lenread) < 0 ) {
				was_error = 1;
				goto done;
			}
			IMA_ADPCM_encoded = 1;
			break;
		case MP3_CODE:
			SDL_SetError("MPEG Layer 3 data not supported",
					SDL_SwapLE16(format->encoding));
			was_error = 1;
			goto done;
		default:
			SDL_SetError("Unknown WAVE data format: 0x%.4x",
					SDL_SwapLE16(format->encoding));
			was_error = 1;
			goto done;
	}
	SDL_memset(spec, 0, (sizeof *spec));
	spec->freq = SDL_SwapLE32(format->frequency);
	switch (SDL_SwapLE16(format->bitspersample)) {
		case 4:
			if ( MS_ADPCM_encoded || IMA_ADPCM_encoded ) {
				spec->format = AUDIO_S16;
			} else {
				was_error = 1;
			}
			break;
		case 8:
			spec->format = AUDIO_U8;
			break;
		case 16:
			spec->format = AUDIO_S16;
			break;
		default:
			was_error = 1;
			break;
	}
	if ( was_error ) {
		SDL_SetError("Unknown %d-bit PCM data format",
			SDL_SwapLE16(format->bitspersample));
		goto done;
	}
	spec->channels = (Uint8)SDL_SwapLE16(format->channels);
	spec->samples = 4096;		/* Good default buffer size */

	/* Read the audio data chunk */
	*audio_buf = NULL;
	do {
		if ( *audio_buf != NULL ) {
			SDL_free(*audio_buf);
			*audio_buf = NULL;
		}
		lenread = ReadChunk(src, &chunk);
		if ( lenread < 0 ) {
			was_error = 1;
			goto done;
		}
		*audio_len = lenread;
		*audio_buf = chunk.data;
		if(chunk.magic != DATA) headerDiff += lenread + 2 * sizeof(Uint32);
	} while ( chunk.magic != DATA );
	headerDiff += 2 * sizeof(Uint32); /* for the data chunk and len */

	if ( MS_ADPCM_encoded ) {
		if ( MS_ADPCM_decode(audio_buf, audio_len) < 0 ) {
			was_error = 1;
			goto done;
		}
	}
	if ( IMA_ADPCM_encoded ) {
		if ( IMA_ADPCM_decode(audio_buf, audio_len) < 0 ) {
			was_error = 1;
			goto done;
		}
	}

	/* Don't return a buffer that isn't a multiple of samplesize */
	samplesize = ((spec->format & 0xFF)/8)*spec->channels;
	*audio_len &= ~(samplesize-1);

done:
	if ( format != NULL ) {
		SDL_free(format);
	}
	if ( src ) {
		if ( freesrc ) {
			SDL_RWclose(src);
		} else {
			/* seek to the end of the file (given by the RIFF chunk) */
			SDL_RWseek(src, wavelen - chunk.length - headerDiff, RW_SEEK_CUR);
		}
	}
	if ( was_error ) {
		spec = NULL;
	}
	return(spec);
}

/* Since the WAV memory is allocated in the shared library, it must also
   be freed here.  (Necessary under Win32, VC++)
 */
void SDL_FreeWAV(Uint8 *audio_buf)
{
	if ( audio_buf != NULL ) {
		SDL_free(audio_buf);
	}
}

static int ReadChunk(SDL_RWops *src, Chunk *chunk)
{
	chunk->magic	= SDL_ReadLE32(src);
	chunk->length	= SDL_ReadLE32(src);
	chunk->data = (Uint8 *)SDL_malloc(chunk->length);
	if ( chunk->data == NULL ) {
		SDL_Error(SDL_ENOMEM);
		return(-1);
	}
	if ( SDL_RWread(src, chunk->data, chunk->length, 1) != 1 ) {
		SDL_Error(SDL_EFREAD);
		SDL_free(chunk->data);
		chunk->data = NULL;
		return(-1);
	}
	return(chunk->length);
}
